app/Http/Controllers/RecruitingCampaignController.php

```diff
@@ -0,0 +1,165 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Log;
+use Notification;
+use Carbon\Carbon;
+use App\RecruitingVisit;
+use App\RecruitingCampaign;
+use Illuminate\Http\Request;
+use App\RecruitingCampaignRecipient;
+use App\Notifications\GeneralInterestNotification;
+
+class RecruitingCampaignController extends Controller
+{
+    public function __construct()
+    {
+        $this->middleware(['permission:send-notifications']);
+    }
+
+    /**
+     * Display a listing of the resource.
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function index()
+    {
+        $rc = RecruitingCampaign::all();
+
+        return response()->json(['status' => 'success', 'campaigns' => $rc]);
+    }
+
+    /**
+     * Store a newly created resource in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\Response
+     */
+    public function store(Request $request)
+    {
+        $this->validate($request, [
+            'name' => 'required|string',
+            'notification_template_id' => 'numeric|exists:notification_templates,id',
+            'start_date' => 'date|required',
+            'end_date' => 'date|required',
+        ]);
+
+        // Store the campaign
+        // Yes, I know there is an easier way to do this.
+        $rc = new RecruitingCampaign();
+        $fields = array_keys($request->all());
+        foreach ($fields as $field) {
+            $rc->$field = $request->input($field);
+        }
+        $rc->created_by = $request->user()->id;
+        $rc->status = 'new';
+
+        try {
+            $rc->save();
+        } catch (QueryException $e) {
+            Bugsnag::notifyException($e);
+            $errorMessage = $e->errorInfo[2];
+
+            return response()->json(['status' => 'error', 'message' => $errorMessage], 500);
+        }
+
+        // Import recipients from visits
+        $start = $request->input('start_date');
+        $end = $request->input('end_date');
+        $visits = RecruitingVisit::where('created_at', '>=', $start)
+                ->where('created_at', '<=', $end)
+                ->get();
+
+        $added_recipient_emails = [];
+        foreach ($visits as $v) {
+            if (in_array($v->recruiting_email, $added_recipient_emails)) {
+                Log::info(get_class().": Email '$v->recruiting_email' already in the list. Ignoring.'");
+            } else {
+                // Add new recipient
+                $rcr = new RecruitingCampaignRecipient();
+                $rcr->email_address = $v->recruiting_email;
+                $rcr->source = 'recruiting_visit';
+                $rcr->recruiting_visit_id = $v->id;
+                $rcr->recruiting_campaign_id = $rc->id;
+                if ($v->user_id != null) {
+                    $rcr->user_id = $v->user_id;
```

Comment:
Why do we copy this

Reply:
the real answer now that I think about this more would be to have made that class (recipients) polymorphic and not copy any of the data

but I didn't think about that until now, so that's why that's there